presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java

```diff
@@ -286,46 +287,73 @@ else if (partitioningHandle.equals(SCALED_WRITER_DISTRIBUTION)) {
             bucketToPartition = Optional.of(new int[1]);
         }
         else {
-            // nodes are pre determined by the nodePartitionMap
-            NodePartitionMap nodePartitionMap = partitioningCache.apply(plan.getFragment().getPartitioning());
-
             Map<PlanNodeId, SplitSource> splitSources = plan.getSplitSources();
             if (!splitSources.isEmpty()) {
-                // contains local source
+                // contains local source, bucket processing
                 List<PlanNodeId> schedulingOrder = plan.getFragment().getPartitionedSources();
                 List<ConnectorPartitionHandle> connectorPartitionHandles;
-                if (plan.getFragment().getStageExecutionStrategy().isAnyScanGroupedExecution()) {
+                boolean groupedExecutionForStage = plan.getFragment().getStageExecutionStrategy().isAnyScanGroupedExecution();
+                if (groupedExecutionForStage) {
                     connectorPartitionHandles = nodePartitioningManager.listPartitionHandles(session, partitioningHandle);
-                    checkState(connectorPartitionHandles.size() == nodePartitionMap.getBucketToPartition().length);
                     checkState(!ImmutableList.of(NOT_PARTITIONED).equals(connectorPartitionHandles));
                 }
                 else {
                     connectorPartitionHandles = ImmutableList.of(NOT_PARTITIONED);
                 }
 
-                int nodeCount = nodePartitionMap.getPartitionToNode().size();
-                ImmutableList.Builder<Node> stageNodeListBuilder = ImmutableList.builderWithExpectedSize(nodeCount);
-                for (int i = 0; i < nodeCount; i++) {
-                    Node node = nodePartitionMap.getPartitionToNode().get(i);
-                    checkState(node != null);
-                    stageNodeListBuilder.add(node);
+                if (plan.getSubStages().isEmpty() // no remote source
+                        && groupedExecutionForStage
+                        // connector has no bucket to node map
+                        && !nodePartitioningManager.getBucketToNode(session, partitioningHandle).isPresent()) {
```

Comment:
I am also thinking to control this by a session property (e.g. `dyanmic-schedule-for-grouped-execution`).

Or, it might be a enum `grouped-execution-lifespan-schedule:
```
FIXED,
DYNAMIC_IF_POSSIBLE,
DYNAMIC_AND_RECOVERABLE_IF_POSSIBLE
```